simpleviewinc/keg-herkin
# Zen 533 save file

**Ticket**: [issue_id](https://jira.simpleviewtools.com/browse/ZEN-533)

## Context

* We need the front-end save action to work

## Goal

* Update the save button on the herkin tap to work
    * it should save it to your local machine
* Also setup pending action and UI to let the users know that their file has been modified and needs to be saved
* do some cleanup on the code editor UI
* be able to retain pending changes when switching files

## Updates

* `repos/shared/models/fileModel.js`
  * no need for modified prop since we now have a filePending store key
* `repos/tap/actions/files/api/loadTestFile.js`
 * should always just load what's live
* `repos/tap/actions/files/api/saveFile.js` 
  * ensure we're returning boolean on success. previously it wasnt returning anything on success save
* `repos/tap/actions/files/local/setActiveFile.js`
  * set active file content to be the new content. this only happens after we save
* `repos/tap/components/codeEditor/codeEditor.js`
  * added surface header on the code editor and allow it to show pending indicator when necessary
* `repos/tap/components/modals/testSelectorModal.js`
  * mostly cleanup and ensuring it loads the file properly
* `repos/tap/containers/rootContainer.js`
  * we don't need to open the sidebar automatically if the modal is showing     
* `repos/tap/hooks/useActiveScreenTab.js`
  * helper hook to get the currently active screen tab
* `repos/tap/hooks/usePendingCallback.js`
  * we are using getStore() here because there's an edge case where this callback gets called on file switch and it has the old fileModel. this ensures we always have the latest fileModel  
## Testing
* pull this branch
* keg herkin start --config ./repos/example/herkin.config.js --from keg-herkin:zen-533-save-file
* navigate to http://herkin-zen-533-save-file.local.kegdev.xyz/
* Select `example.feature` as the initial file and `Editor` as the screen
   * [image](https://user-images.githubusercontent.com/3317835/110013550-fe8a6a00-7cde-11eb-81a9-9bb50259ba67.png)
* Verify that the Editor loads with the file content on the left editor
* Verify that the surface header is `Editor - Feature` 
  * [image](https://user-images.githubusercontent.com/3317835/110013711-3396bc80-7cdf-11eb-9493-1b13d71172e9.png)
* make changes to the feature file
* Verify that pending indicator is shows on the surface header and side menu
  * [image](https://user-images.githubusercontent.com/3317835/110015732-735ea380-7ce1-11eb-891b-dc9c8b7aa596.png)
* click the `Save` btn
* the pending indicator should dissapear and you should see the file saved to your local
  * [image](https://user-images.githubusercontent.com/3317835/110015847-97ba8000-7ce1-11eb-9798-566f10f26822.png)
* Verify that if you modify several files without saving. the pending content persist on file switch
 * [image](https://user-images.githubusercontent.com/3317835/110016004-c5072e00-7ce1-11eb-9e53-9fe65ca87d67.png)
